File: leetcode/1404/dotnet/Solution/Solution.cs
using System.Collections;

namespace Solution;

public class Solution
{
    public int NumSteps(string s)
    {
        int offset = 1;
        int steps = 0;
        
        BitArray bitArray = this.InitializeBitArray(s, offset);
        
        while (bitArray.Count - offset > 1)
        {
            steps++;
            if (this.IsOdd(bitArray))
                offset = this.SumOne(bitArray, offset);
            else
            {
                offset = this.DivideByTwo(bitArray, offset);
            }
        }
        return steps;
    }

    private int DivideByTwo(BitArray bitArray, int offset)
    {
        bitArray.LeftShift(1);
        return ++offset;
    }

    private BitArray InitializeBitArray(string s, int offset)
    {
        var bitArray = new BitArray(s.Length + offset);
        for (var i = 0; i < s.Length; i++)
        {
            bitArray.Set(i + offset, s[i] == '1');
        }

        return bitArray;
    }

    private bool IsOdd(BitArray bitArray)
    {
        return bitArray.Get(bitArray.Count - 1);
    }
    private int SumOne(BitArray bitArray, int offset)
    {

        for (var i = bitArray.Count - 1; i >= 0 && i >= offset - 1; i--)
        {
            if (i < offset)
                offset = i;
            var currentPosition = bitArray.Get(i);
            bitArray.Set(i, !currentPosition);
            if (!currentPosition)
                break;
        }
        return offset;
    }
}

File: leetcode/1404/go/solution/solution.go
package solution

func numSteps(s string) int {
	offset := 1
	steps := 0

	bitArray := make([]bool, len(s)+offset)
	for i := 0; i < len(s); i++ {
		bitArray[i+offset] = s[i] == '1'
	}

	for len(bitArray)-offset > 1 {
		steps++
		if isOdd(&bitArray) {
			offset = sumOne(&bitArray, offset)

		} else {
			offset = divideByTwo(&bitArray, offset)
		}
	}
	return steps
}

func isOdd(bitArray *[]bool) bool {
	return (*bitArray)[len(*bitArray)-1]
}

func sumOne(bitArray *[]bool, offset int) int {
	for i := len(*bitArray) - 1; i >= 0 && i >= offset-1; i-- {
		if i < offset {
			offset = i
		}
		currentPosition := (*bitArray)[i]
		(*bitArray)[i] = !currentPosition
		if !currentPosition {
			break
		}
	}
	return offset
}

func divideByTwo(bitArray *[]bool, offset int) int {
	*bitArray = (*bitArray)[:len(*bitArray)-1]
	return offset
}
